### PR TITLE
[CELEBORN-1648] Refine AppUniqueId with UUID suffix

### PR DESCRIPTION
**Pull request by chenkovsky**
### What changes were proposed in this pull request?

We can add randomUUID as an suffix to solve it

### Why are the changes needed?

currently, we cannot guarantee application id is really unique. this may lead to data issue.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?
test locally

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/util/Utils.scala [hunk 1]
```diff
@@ -1305,4 +1305,8 @@ object Utils extends Logging {
         (key, value)
     }.asInstanceOf[Seq[(K, V)]]
   }
+
+  def appUniqueIdWithUUID(appUiqueId: String): String = {
+    appUiqueId + "-" + UUID.randomUUID()
```
Comment (SteNicholas) [inline common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:1310]: Could you add config to allow user to add UUID suffix? IMO, not all cases should add UUID suffix. Right?

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java [hunk 2]
```diff
@@ -76,7 +77,9 @@ public void registerJob(JobShuffleContext context) {
     if (lifecycleManager == null) {
       synchronized (RemoteShuffleMaster.class) {
         if (lifecycleManager == null) {
-          celebornAppId = FlinkUtils.toCelebornAppId(lifecycleManagerTimestamp, jobID);
+          celebornAppId =
+              Utils.appUniqueIdWithUUID(
```
[the remark this answers is not in the document]
Comment (chenkovsky) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java:81]: so you think there's no application id colision problem for flink?

[the remark this answers is not in the document]
Comment (SteNicholas) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java:81]: Yes. cc @reswqa, @RexXiong.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -5055,6 +5056,14 @@ object CelebornConf extends Logging {
       .checkValue(v => v > 0.0 && v <= 1.0, "Value must be between 0 and 1 (inclusive)")
       .createWithDefault(0.4)
 
+  val CLIENT_APPLICATION_UUID: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.application.uuid.enabled")
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:5060]: perfer celeborn.client.application.uuidSuffix.enabled

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 4]
```diff
@@ -905,6 +905,7 @@ class CelebornConf(loadDefaults: Boolean) extends Cloneable with Logging with Se
   def clientExcludeReplicaOnFailureEnabled: Boolean =
     get(CLIENT_EXCLUDE_PEER_WORKER_ON_FAILURE_ENABLED)
   def clientMrMaxPushData: Long = get(CLIENT_MR_PUSH_DATA_MAX)
+  def clientApplicationUUID: Boolean = get(CLIENT_APPLICATION_UUID)
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:908]: clientApplicationUUID -> clientApplicationUUIDSuffixEnabled

File: common/src/main/scala/org/apache/celeborn/common/util/Utils.scala (hunk 1 above)
Comment (RexXiong) [reply common/src/main/scala/org/apache/celeborn/common/util/Utils.scala:1310]: better to move appUniqueIdWithUUID(appUiqueId: String) to CelebornConf. Additionally, we can simplify the UUID by removing the dashes to create a shorter version, then we can use celebornConf.appUniqueIdWithUUIDSuffix

```suggestion
  def appUniqueIdWithUUIDSuffix(appId: String): String = {
     if (clientApplicationUUIDSuffixEnabled) {
        appId + "-" + UUID.randomUUID().toString().replaceAll("-", "")
     } else {
       appId
     }
   }

```

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (RexXiong) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java]: > Yes. cc @reswqa, @RexXiong.

Currently Flink Appid is already random and concat with current timestamp,  but I think it's fine if we add flag for UUID suffix

[the remark this answers is not in the document]
Comment (reswqa) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java]: I didn't expect that flink app id(JobId + timestamp) might conflict. Did you actually suffer from this problem.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java [hunk 5]
```diff
@@ -76,12 +77,17 @@ public void registerJob(JobShuffleContext context) {
...
       synchronized (RemoteShuffleMaster.class) {
         if (lifecycleManager == null) {
-          celebornAppId = FlinkUtils.toCelebornAppId(lifecycleManagerTimestamp, jobID);
-          LOG.info("CelebornAppId: {}", celebornAppId);
+          String applicationId = FlinkUtils.toCelebornAppId(lifecycleManagerTimestamp, jobID);
           CelebornConf celebornConf =
               FlinkUtils.toCelebornConf(shuffleMasterContext.getConfiguration());
           // if not set, set to true as default for flink
           celebornConf.setIfMissing(CelebornConf.CLIENT_CHECKED_USE_ALLOCATED_WORKERS(), true);
+          if (celebornConf.clientApplicationUUID()) {
```
Comment (SteNicholas) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java:85]: Could you add util method for `clientAppId` in `FlinkUtils`?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 6]
```diff
@@ -5055,6 +5056,14 @@ object CelebornConf extends Logging {
       .checkValue(v => v > 0.0 && v <= 1.0, "Value must be between 0 and 1 (inclusive)")
       .createWithDefault(0.4)
 
+  val CLIENT_APPLICATION_UUID: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.application.uuid.enabled")
+      .categories("client")
+      .version("0.6.0")
+      .doc("When `true`, add uuid suffix to application id")
```
Comment (SteNicholas) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:5063]:
```suggestion
      .doc("Whether to add UUID suffix for application id for unique. When `true`, add UUID suffix for unique application id.")
```

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java [hunk 7]
```diff
@@ -76,12 +77,17 @@ public void registerJob(JobShuffleContext context) {
     if (lifecycleManager == null) {
       synchronized (RemoteShuffleMaster.class) {
         if (lifecycleManager == null) {
-          celebornAppId = FlinkUtils.toCelebornAppId(lifecycleManagerTimestamp, jobID);
-          LOG.info("CelebornAppId: {}", celebornAppId);
+          String applicationId = FlinkUtils.toCelebornAppId(lifecycleManagerTimestamp, jobID);
```
Comment (SteNicholas) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java:80]: Please revert this change. Meanwhile, remove line 87~89.

File: client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierMasterAgent.java [hunk 8]
```diff
@@ -85,7 +86,12 @@ public void registerJob(JobID jobID, TierShuffleHandler tierShuffleHandler) {
     if (lifecycleManager == null) {
       synchronized (CelebornTierMasterAgent.class) {
         if (lifecycleManager == null) {
-          celebornAppId = FlinkUtils.toCelebornAppId(lifecycleManagerTimestamp, jobID);
+          String applicationId = FlinkUtils.toCelebornAppId(lifecycleManagerTimestamp, jobID);
```
Comment (SteNicholas) [inline client-flink/flink-1.20/src/main/java/org/apache/celeborn/plugin/flink/tiered/CelebornTierMasterAgent.java:89]: Ditto.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java (hunk 7 above)
Comment (chenkovsky) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java:80]: > Please revert this change. Meanwhile, remove line 87~89.

ok, let's revert it.
should we describe that the flag is only for spark and mr in doc ?

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (chenkovsky) [reply client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleMaster.java]: @SteNicholas please review it again, I reverted the change.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 9]
```diff
@@ -5055,6 +5064,14 @@ object CelebornConf extends Logging {
       .checkValue(v => v > 0.0 && v <= 1.0, "Value must be between 0 and 1 (inclusive)")
       .createWithDefault(0.4)
 
+  val CLIENT_APPLICATION_UUID_SUFFIX_ENABLED: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.application.uuidSuffix.enabled")
+      .categories("client")
+      .version("0.6.0")
+      .doc("Whether to add UUID suffix for application id for unique. When `true`, add UUID suffix for unique application id.")
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:5071]: This configuration only takes effect when using the Spark and MR engines, so we must specify this in doc.

**Review by RexXiong — APPROVED** [review]
LGTM, only a minor comment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 10]
```diff
@@ -5055,6 +5064,14 @@ object CelebornConf extends Logging {
       .checkValue(v => v > 0.0 && v <= 1.0, "Value must be between 0 and 1 (inclusive)")
       .createWithDefault(0.4)
 
+  val CLIENT_APPLICATION_UUID_SUFFIX_ENABLED: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.application.uuidSuffix.enabled")
+      .categories("client")
+      .version("0.6.0")
+      .doc("Whether to add UUID suffix for application id for unique. When `true`, add UUID suffix for unique application id. Currently, this only applies to Spark and MapPartition.")
```
Comment (RexXiong) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:5071]: MapPartition?

**Review by RexXiong — APPROVED** [review]

**Comment by RexXiong** [comment]
Thanks, merge to main(v0.6.0)

**Comment by chenkovsky** [comment]
> Circling back to this - is this a case where multiple application attempts were involved ? (in which case, using appId + attemptId is the right approach) - or when different application;s application id actually collided?
> 
> Given spark (and MR ?) use application id + attempt id for a few things already (like event files), I am trying to understand why UUID was necessary.

Hi, @mridulm . In our cluster, most of time, applicationId is unique, but id collision will take place occasionally. when it happens, it's very hard to find the root cause of data missing problem. For spark or MR, user can also override appId, add uuid suffix is also a fool-proof design.
